### DISCUSSION TITLE
Middleware to write access log with request and response body and other data

### DISCUSSION THREAD

**Discussion by VatsalJagani** [fastapi/fastapi#11330] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@app.middleware('http')
async def access_log(request: Request, call_next):
    """Middleware to log request details and response."""
    start_time = time.time()

    client_ip = request.client.host
    method = request.method
    path = request.url.path
    headers = dict(request.headers)
    params = dict(request.query_params)

    # Handle request body based on content type
    request_body = None
    try:
        request_body = await request.body()
        if request_body:
            request_body = request_body.decode("utf-8")
            if request_body:
                # Truncate large bodies to avoid excessive logging
                request_body = request_body[:1024]  # Adjust truncation limit as needed
                request_body += " <Body-Truncated>"
        logger.info(f"Trying to get request body again {await request.body()}")
    except Exception as e:
        request_body = f"Error reading body: {str(e)}"

    try:
        response = await call_next(request)
    except HTTPException as e:
        response = JSONResponse(content={"error": str(e)}, status_code=e.status_code)
        exception_details = f"HTTPException: {str(e)}"
    except Exception as e:
        response = JSONResponse(content={"error": "Internal Server Error"}, status_code=500)
        exception_details = f"Exception: {str(e)}"
    else:
        exception_details = None

    process_time = time.time() - start_time

    response_body = None
    try:
        response_body = b""
        async for chunk in response.body_iterator:
            response_body += chunk

        if response_body:
            response_body = response_body.decode("utf-8")
            if len(response_body) > 1024:
                # Truncate large bodies to avoid excessive logging
                response_body = response_body[:1024]  # Adjust truncation limit as needed
                response_body += " <Body-Truncated>"
        logger.info(f"Trying to get response body again {await response.body()}")
    except Exception as e:
        logger.error(f"Error getting response body. {e}")

    log_data = {
        "client_ip": client_ip,
        "method": method,
        "path": path,
        "headers": headers,
        "request_params": params,
        "request_body": request_body,
        "response_status": response.status_code,
        "response_body": response_body,
        "exception_details": exception_details,
        "process_time": process_time,
    }

    logger.info(f"API request: {log_data}")

    return response
```

### Description

The given code is to add middleware to write the complete access log into a single line to better log management.

Here I have two issues with accessing the request and response body. Once read it cleans the body. I have already tried this (https://github.com/tiangolo/fastapi/issues/394#issuecomment-883524819) for request body management. It used to work but now it is giving me an error somehow.

```
    message = await receive()
  File "C:\Users\vatsal\Documents\My_Projects\Invitelly-Backend\.venv\lib\site-packages\starlette\middleware\base.py", line 56, in wrapped_receive
    raise RuntimeError(f"Unexpected message received: {msg['type']}")
RuntimeError: Unexpected message received: http.request
```

And I could not yet find a solution to log the **response body**.

I found a solution around wrapping the request and response object to another request and response class, but I don't think that would make exact replica of the original object, so I'm afraid to use that.

### Operating System

Windows

### Operating System Details

This issue it not related to the operating system.

### FastAPI Version

0.110.0

### Pydantic Version

2.6.4

### Python Version

3.10.5

**Comment by YuriiMotov** [comment]
Let's close this for now to keep the discussions clean and organized.
If you’ve found a solution, please share it here.
If you still need help, feel free to let us know.
